bindings/go/input/dir/blob.go

```diff
@@ -0,0 +1,210 @@
+	fs, err := filesystem.NewFS(baseDir, os.O_RDONLY)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create filesystem while trying to access %v: %w", path, err)
+	}
+
+	// Create a new tar writer.
+	var buf bytes.Buffer
+	tw := tar.NewWriter(&buf)
+	defer func() {
+		closeErr := tw.Close()
+		if closeErr != nil && err == nil {
+			err = closeErr
+		}
+	}()
+
+	// Walk recursively through directory contents and add it to the tar.
+	err = walkDirContents(path, baseDir, dir.ExcludeFiles, dir.IncludeFiles, fs, tw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to package directory contents as a tar archive: %w", err)
+	}
+
+	// Close the tar writer
+	if err := tw.Close(); err != nil {
+		return nil, fmt.Errorf("failed to close tar writer: %w", err)
+	}
+
+	return bytes.NewReader(buf.Bytes()), nil
+}
+
+// walkDirContents does recursive packaging of the directory contents.
+// The function goes the directory contents file by file, checks if it should be included or excluded,
+// creates tar headers for each file and subfolder, and writes the file contents to the tar archive.
+// For subdirectories it calls itself recursively to process the subfolder contents.
+func walkDirContents(currentDir string, baseDir string,
+	excludePatterns, includePatterns []string,
+	fs filesystem.FileSystem, tw *tar.Writer,
+) (err error) {
+	// Read directory contents.
+	dirRelPath, err := filepath.Rel(baseDir, currentDir)
+	if err != nil {
+		return fmt.Errorf("failed to get directory relative path: %w", err)
+	}
+	dirEntries, err := fs.ReadDir(dirRelPath)
+	if err != nil {
+		return fmt.Errorf("failed to read directory entries: %w", err)
+	}
+
+	// Iterate over directory entries.
+	for _, entry := range dirEntries {
+		// Get FileInfo for the entry.
+		info, err := entry.Info()
+		if err != nil {
+			return fmt.Errorf("failed to get file information: %w", err)
+		}
+
+		// Construct the full path of the entry.
+		entryPath := filepath.Join(currentDir, entry.Name())
+		// Get the relative path of the entry with respect to the base directory.
+		relPath, err := filepath.Rel(baseDir, entryPath)
+		if err != nil {
+			return fmt.Errorf("failed to get relative path: %w", err)
+		}
+		// Check, if the entry should be included in the tar archive.
+		include, err := isPathIncluded(relPath, excludePatterns, includePatterns)
+		if err != nil {
+			return fmt.Errorf("failed to check if an entry should be included in the tar archive: %w", err)
+		}
+		if !include {
+			continue
+		}
+
+		// Create tar header.
+		header, err := tar.FileInfoHeader(info, "")
+		if err != nil {
+			return fmt.Errorf("failed to create tar header: %w", err)
+		}
+		header.Name = relPath
+
+		// Write the header to the tar archive.
+		if err := tw.WriteHeader(header); err != nil {
+			return fmt.Errorf("failed to write tar header to tar archive: %w", err)
+		}
+
+		// If the entry is a file, copy its content to the tar archive.
+		if entry.Type().IsRegular() {
```

Comment:
what if the entrz is neither regular nor a directory? maybe we should error or at least warn the user?